docs/java/how-to-develop-a-workflow-definition-in-java.md

```diff
@@ -0,0 +1,186 @@
+    List<String> getHistory();
+
+    @QueryMethod
+    String getStatus();
+
+    @SignalMethod
+    void abandon();
+}
+```
+
+With this, another Workflow interface can extend just `Retryable`, for example:
+
+```java
+@WorkflowInterface
+public interface MediaProcessingWorkflow extends Retryable {
+
+    @WorkflowMethod
+    String processBlob(Arguments args);
+}
+```
+
+Now if we have two running Workflows, one that implements the `FileProcessingWorkflow` interface and another that implements the
+`MediaProcessingWorkflow` interface, we can send a Signal to both using their common interface and their respective workflowIds, for example:
+
+```java
+Retryable r1 = client.newWorkflowStub(Retryable.class, firstWorkflowId);
+Retryable r2 = client.newWorkflowStub(Retryable.class, secondWorkflowId);
+r1.retryNow();
+r2.retryNow();
+```
+
+The same technique can be used to send a Query to the Workflows using a base Workflow interface.
+
+Note that this approach does not apply to `@WorkflowMethod` annotations, meaning that when using a base interface, it should
+not include any `@WorkflowMethod` methods.
+To illustrate this, lets' say that we define the following **invalid** code:
+
+```java
+// INVALID CODE!
+public interface BaseWorkflow {
+    @WorkflowMethod
+    void retryNow();
+}
+
+@WorkflowInterface
+public interface Workflow1 extends BaseWorkflow {}
+
+@WorkflowInterface
+public interface Workflow2 extends BaseWorkflow {}
+```
+
+If we attempt to register implementations of Workflow1 and Workflow2 with a Worker, it will fail. Let's say that we have:
+
+```java
+worker.registerWorkflowImplementationTypes(
+        Workflow1Impl.class, Workflow2Impl.class);
+```
+
+This registration will fail with:
+
+```text
+java.lang.IllegalStateException: BaseWorkflow workflow type is already registered with the worker
+```
+
+## Implementing Workflows
+
+A Workflow implementation implements a Workflow interface. Each time a new Workflow Execution is started,
+a new instance of the Workflow implementation object is created.
+Then, one of the methods (depending on which Workflow type has been started) annotated with `@WorkflowMethod` can be invoked.
+As soon as this method returns, the Workflow execution is considered as completed.
+
+Workflow methods annotated with `@QueryMethod` and `@SignalMethod` can be invoked during a Workflow's execution.
+
+Note that methods annotated with `@QueryMethod` can be invoked even when a Workflow is in the `Completed`
+state.
+
+### Workflow Implementation Constraints
+
+Temporal uses the [Event Sourcing pattern](https://docs.microsoft.com/en-us/azure/architecture/patterns/event-sourcing) to recover
+the state of a Workflow object including its threads and local variable values.
+In essence, every time a Workflow state has to be restored, its code is re-executed from the beginning.
+Note that during replay, successfully executed Activities are not re-executed as their results are already recorded
```

Comment:
```suggestion
Note that, during replay, successfully executed Activities are not re-executed because their results are already recorded in the Event History of the Workflow Execution.
```